include/envoy/network/filter.h

```diff
@@ -232,6 +234,47 @@ class ListenerFilterManager {
  */
 typedef std::function<void(ListenerFilterManager& filter_manager)> ListenerFilterFactoryCb;
 
+/**
+ * Interface representing a single filter chain.
+ */
+class FilterChain {
+public:
+  virtual ~FilterChain() {}
+
+  /**
+   * @return bool true if filter chain implements secure transport, false if it doesn't.
+   */
+  virtual bool implementsSecureTransport() const PURE;
+
+  /**
+   * @return Network::TransportSocketPtr a transport socket to be used by the new connection.
+   */
+  virtual TransportSocketPtr createTransportSocket() const PURE;
+
+  /**
+   * const std::vector<FilterFactoryCb>& a list of filters to be used by the new connection.
+   */
+  virtual const std::vector<FilterFactoryCb>& getNetworkFilterFactories() const PURE;
+};
+
+typedef std::shared_ptr<FilterChain> FilterChainSharedPtr;
+
+/**
+ * Interface for searching through configured filter chains.
+ */
+class FilterChainManager {
+public:
+  virtual ~FilterChainManager() {}
+
+  /**
+   * Find filter chain that's matching metadata from the new connection.
+   * @param socket supplies connection metadata that's going to be used for the filter chain lookup.
+   * @return const FilterChainSharedPtr filter chain to be used by the new connection,
+   *         nullptr if no matching filter chain was found.
+   */
+  virtual const FilterChain* findFilterChain(const ConnectionSocket* socket) const PURE;
```

Comment:
nit: Sorry on the param stuff, but can it be `const ConnectionSocket& socket`? It can't be null right?

Reply:
Done.